audio/src/passthrough_decoder.rs

```diff
@@ -0,0 +1,210 @@
+        pck = rdr.read_packet_expected()?;
+    }
+
+    *stream_serial = pck.stream_serial();
+
+    let pkt_type = pck.data[0];
+    debug!("Vorbis header type{}", &pkt_type);
+
+    // all headers are mandatory
+    if pkt_type != code {
+        return Err(PassthroughError::NotVorbisError);
+    }
+
+    let absgp_page = pck.absgp_page();
+    wtr.write_packet(
+        pck.data.into_boxed_slice(),
+        *stream_serial,
+        info,
+        absgp_page,
+    )?;
+
+    return Ok(absgp_page);
+}
+
+static ABSGP_PAGE: AtomicU64 = AtomicU64::new(0);
+
+pub struct PassthroughDecoder<R: Read + Seek> {
+    rdr: PacketReader<R>,
+    wtr: PacketWriter<Vec<u8>>,
+    absgp_page: u64,
+}
+
+impl<R: Read + Seek> PassthroughDecoder<R> {
+    /// Constructs a new Decoder from a given implementation of `Read + Seek`.
+    pub fn new(rdr: R) -> Result<Self, PassthroughError> {
+        let mut rdr = PacketReader::new(rdr);
+        let mut wtr = PacketWriter::new(Vec::new());
+
+        let (_stream_serial, absgp_page) = write_headers(&mut rdr, &mut wtr)?;
+        info!("Starting passthrough track (gp => {})", absgp_page);
+
+        return Ok(PassthroughDecoder {
+            rdr,
+            wtr,
+            absgp_page,
+        });
+    }
+
+    /// Seeks to the specified absolute granule position, with a page granularity.
+    pub fn seek_absgp_pg(&mut self, absgp: u64) -> Result<(), PassthroughError> {
+        // set need to acquire granule
+        self.absgp_page = u64::MAX;
+        self.rdr.seek_absgp(None, absgp)?;
+        Ok(())
+    }
+}
+
+impl<R: Read + Seek> AudioDecoder for PassthroughDecoder<R> {
+    fn seek(&mut self, ms: i64) -> Result<(), AudioError> {
+        info!("Seeking to {}", ms);
+
+        // hard-coded to 44.1 kHz
+        let absgp = ms * 44100 / 1000;
+        self.seek_absgp_pg(absgp as u64)?;
+
+        Ok(())
+    }
+
+    fn next_packet(&mut self) -> Result<Option<AudioPacket>, AudioError> {
+        loop {
+            let pck = match self.rdr.read_packet() {
+                Ok(Some(pck)) => pck,
+
+                Ok(None) | Err(OggReadError::NoCapturePatternFound) => {
+                    info!("end of streaming");
+                    return Ok(None);
+                }
+
+                Err(err) => return Err(err.into()),
+            };
+
+            // then do normal packet processing
+            let inf = if pck.last_in_stream() {
+                PacketWriteEndInfo::EndStream
+            } else if pck.last_in_page() {
+                PacketWriteEndInfo::EndPage
+            } else {
+                PacketWriteEndInfo::NormalPacket
+            };
+
+            // NB: we don't handle multiple streams (change of serial)
+            let stream_serial = pck.stream_serial();
+            let absgp_page = pck.absgp_page();
+
+            // set granule if unknown
+            if self.absgp_page == u64::MAX {
+                self.absgp_page = absgp_page;
+            }
```

Comment:
Feels a bit hacky. Maybe `Option<u64>` would be clearer? And maybe rename the field to `last_absgp_page`?

Reply:
Done but what do you expect from a C developer :-) ?